rpcs3/rpcs3qt/gui_settings.cpp

```diff
@@ -165,34 +168,49 @@ void gui_settings::ShowBox(QMessageBox::Icon icon, const QString& title, const Q
 		return;
 	}
 
-	const QFlags<QMessageBox::StandardButton> buttons = icon != QMessageBox::Information ? QMessageBox::Yes | QMessageBox::No : QMessageBox::Ok;
-
-	QMessageBox* mb = new QMessageBox(icon, title, text, buttons, parent, Qt::Dialog | Qt::MSWindowsFixedSizeDialogHint | (always_on_top ? Qt::WindowStaysOnTopHint : Qt::Widget));
-	mb->deleteLater();
+	QMessageBox* mb = new QMessageBox(icon, title, text, {}, parent, Qt::Dialog | Qt::MSWindowsFixedSizeDialogHint | (always_on_top ? Qt::WindowStaysOnTopHint : Qt::Widget));
 	mb->setTextFormat(Qt::RichText);
 
 	if (has_gui_setting && icon != QMessageBox::Critical)
 	{
 		mb->setCheckBox(new QCheckBox(tr("Don't show again")));
 	}
 
-	connect(mb, &QMessageBox::finished, [&](int res)
-	{
-		if (result)
-		{
-			*result = res;
-		}
+	QAbstractButton* on_install = nullptr; 
+	if (entry == gui::ib_confirm_fw) on_install = mb->addButton(tr("Install"), QMessageBox::ActionRole);
 
-		const auto checkBox = mb->checkBox();
+	QAbstractButton* on_ignore = nullptr;
+	if (entry == gui::ib_confirm_fw) on_ignore = mb->addButton(tr("Ignore"), QMessageBox::ActionRole);
 
-		if (checkBox && checkBox->isChecked())
-		{
-			SetValue(entry, false);
-			cfg_log.notice("%s Dialog for Entry %s is now disabled", dialog_type, sstr(entry.name));
-		}
-	});
+	QAbstractButton* on_okay = nullptr;
+	if (!on_install) on_okay = mb->addButton(tr("Okay"), QMessageBox::ActionRole);
+
+	QAbstractButton* on_cancel = mb->addButton(tr("Cancel"), QMessageBox::ActionRole);
 
 	mb->exec();
+
+	QAbstractButton* clicked = mb->clickedButton();
+
+	if (result)
+	{
+		*result = (clicked && (clicked == on_okay || clicked == on_install || clicked == on_ignore) ? QMessageBox::Yes : QMessageBox::No);
+	}
+
+	const auto checkBox = mb->checkBox();
+
+	if (checkBox && checkBox->isChecked())
+	{
+		SetValue(entry, false);
+		cfg_log.notice("%s Dialog for Entry %s is now disabled", dialog_type, sstr(entry.name));
+	}
+
+	if (clicked && clicked == on_install)
+	{
+		// Special path for firmware installation, parent must be main_window
+		dynamic_cast<main_window*>(parent)->InstallPup();
```

Comment:
Why dynamic_cast?

Reply:
Avoiding UB when misused, ensuring nullptr derference.

Reply:
Which is a plain UB... Perhaps you could compare parent with known pointer instead.

Reply:
Used emu callbacks instead.